daemon/plugin_p.cpp

```diff
@@ -6,43 +6,99 @@
  */
 
 #include "plugin_p.h"
+#include <QtCore/QCoreApplication>
+#include <QtCore/QDir>
 
 using namespace Sailfish::Secrets;
 
-Daemon::ApiImpl::PluginHelper::PluginHelper(
-        const QString &fileName, bool autotestMode)
-    : QPluginLoader(fileName)
-    , m_failureType(NoFailure)
-    , m_autotestMode(autotestMode)
+static Daemon::ApiImpl::PluginManager *pluginManagerInstance = Q_NULLPTR;
+
+static bool isAutotestMode()
 {
+    auto &app = *QCoreApplication::instance();
+    auto args = app.arguments();
+    bool autotestMode = false;
+
+    if (args.size() > 1 &&
+        (args[1] == QLatin1String("test") ||
+         args[1] == QLatin1String("-test") ||
+         args[1] == QLatin1String("--test"))) {
+        autotestMode = true;
+    }
+
+    return autotestMode;
 }
 
-Daemon::ApiImpl::PluginHelper::~PluginHelper()
+Daemon::ApiImpl::PluginManager *Daemon::ApiImpl::PluginManager::instance()
 {
+    if (!pluginManagerInstance) {
+        pluginManagerInstance = new PluginManager();
+    }
+
+    return pluginManagerInstance;
 }
 
-Daemon::ApiImpl::PluginHelper::FailureType
-Daemon::ApiImpl::PluginHelper::failureType() const
+QVector<QPluginLoader*> Daemon::ApiImpl::PluginManager::loadPluginFiles()
 {
-    return m_failureType;
+    QVector<QPluginLoader*> result;
+    QStringList paths = QCoreApplication::libraryPaths();
+
+    Q_FOREACH(const QString &path, paths) {
+        // Don't enumerate /usr/bin
+        if (path == "/usr/bin")
+            continue;
+
+        QDir dir(path);
+        Q_FOREACH (const QFileInfo &file, dir.entryInfoList(QDir::Files | QDir::NoDot | QDir::NoDotDot, QDir::Name)) {
+            const QString fileName = file.fileName();
+
+            // Don't even try to load files which don't look like libraries
+            if (!fileName.startsWith("lib") || !fileName.contains(".so")) {
+                continue;
+            }
+
+            // load the plugin
+            auto *loader = new QPluginLoader(file.absoluteFilePath());
+            if (!loader->load()) {
+                qWarning() << "Could not load plugin:" << loader->fileName();
+                delete loader;
+                continue;
+            }
+
+            result.append(loader);
+        }
+    }
+
+    return result;
 }
 
-void Daemon::ApiImpl::PluginHelper::reportFailure(
-        const QLoggingCategory &category())
+void Daemon::ApiImpl::PluginManager::addPlugin(QPluginLoader *loader, const PluginHelpers::PluginInfo &info, QObject *obj)
 {
-    if (m_failureType != PluginLoadFailure
-            && m_failureType != PluginTypeFailure) {
-        // already reported this error
-        return;
+    bool use = true;
+    bool testMode = isAutotestMode();
+
+    if (!info.canUse) {
+        qWarning() << "Not a useable crypto or secrets plugin:" << loader->fileName();
+        use = false;
+    }
+
+    if (m_plugins.contains(info.name)) {
+        qWarning() << "Not adding plugin with duplicate name:" << loader->fileName();
```

Comment:
I think we should use qCWarning() with an appropriate logging category instead of raw qWarning

Reply:
Which category should it be?

Reply:
I'd suggest something like "sailfishsecrets.plugin-loader" ?

Reply:
Ok.

Reply:
Currently we have `org.sailfishos.crypto.daemon` and `org.sailfishos.secrets.daemon`.